File: TypeScript/Algorithm/Algorithm I/977. Squares of a Sorted Array.ts
/**
 * Squares each element of a sorted array and returns a new array sorted in non-decreasing order.
 * @param {number[]} nums - The sorted array of integers.
 * @returns {number[]} A new sorted array with each element squared.
 */
function sortedSquares(nums: number[]): number[] {
  const numsLength: number = nums.length;
  const squareSortArray: number[] = new Array(numsLength);

  let leftPointer: number = 0;
  let rightPointer: number = numsLength - 1;

  for (let index: number = numsLength - 1; index >= 0; index--) {
    const squareStart: number = nums[leftPointer] ** 2;
    const squareEnd: number = nums[rightPointer] ** 2;

    squareSortArray[index] = squareEnd > squareStart ? squareEnd : squareStart;

    if (squareEnd > squareStart) rightPointer--;
    else leftPointer++;
  }

  return squareSortArray;
}

File: Java/Algorithm/Algorithm I/977. Squares of a Sorted Array.java
class Solution {

  public int[] sortedSquares(int[] nums) {
    int NUMS_LENGTH = nums.length;

    int[] sortedSquareArray = new int[NUMS_LENGTH];
    int indexInsertion = NUMS_LENGTH - 1;

    int leftPointer = 0;
    int rightPointer = NUMS_LENGTH - 1;

    while (leftPointer <= rightPointer) {
      int firstValue = Math.abs(nums[leftPointer]);
      int lastValue = Math.abs(nums[rightPointer]);

      if (firstValue > lastValue) {
        sortedSquareArray[indexInsertion] = firstValue * firstValue;
        leftPointer++;
      } 
      
      else {
        sortedSquareArray[indexInsertion] = lastValue * lastValue;
        rightPointer--;
      }

      indexInsertion--;
    }

    return sortedSquareArray;
  }
}
